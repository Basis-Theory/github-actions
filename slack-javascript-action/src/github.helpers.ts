const github = require("@actions/github");
const core = require("@actions/core"); // For managing action inputs/outputs

export async function getJobStatuses() {
  const token = process.env.GITHUB_TOKEN; // Automatically passed to your action
  console.log("token exists", token?.length);
  if (!token) {
    throw new Error("GITHUB_TOKEN not found in environment");
  }
  const octokit = github.getOctokit(token);

  const { owner, repo } = github.context.repo;
  const runId = github.context.runId; // The workflow run ID

  try {
    // Fetch the jobs for the specific workflow run
    const { data: jobs } = await octokit.actions.listJobsForWorkflowRun({
      owner,
      repo,
      run_id: runId,
    });

    // Loop through jobs and print job statuses
    jobs.jobs.forEach((job: any) => {
      console.log(
        `Job: ${job.name}, Status: ${job.status}, Conclusion: ${job.conclusion}`
      );
    });

    // Check if any job was canceled
    const cancelledJobs = jobs.jobs.filter(
      (job: any) => job.status === "cancelled"
    );
    if (cancelledJobs.length > 0) {
      console.log("There were canceled jobs:", cancelledJobs);
    } else {
      console.log("No jobs were canceled.");
    }
  } catch (error) {
    console.error("Error fetching job statuses:", error);
  }
}
