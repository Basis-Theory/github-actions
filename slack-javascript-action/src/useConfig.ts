import { getInput, exportVariable } from "@actions/core";

export type ConfigType = {
  repository: string;
  version: string;
  release_notes: string;
  author: string;
  type: string;
  action_url: string;
  message_id: string | undefined;
  startedTimestamp: string;
  stoppedTimestamp: string;
  status: string;
  job_status: string | undefined;
  channel: string;
  mention_person: string | undefined;
};

export type GithubContextCommitType = {
  message: string;
};

export type GithubContextType = {
  actor: string;
  sha: string;
  server_url: string;
  repository: string;
  run_id: string;
  event: {
    commits?: GithubContextCommitType[];
    repository: {
      name: string;
    };
    release?: {
      tag_name: string;
      body: string;
      author: {
        login: string;
      };
    };
  };
};

const getVersion = (githubContext: GithubContextType): string =>
  githubContext.event.release
    ? githubContext.event.release.tag_name
    : githubContext.sha.slice(0, 6);

const getReleaseNotes = (githubContext: GithubContextType): string => {
  if (githubContext.event.release) {
    return githubContext.event.release.body;
  }

  if (githubContext.event.commits) {
    return githubContext.event.commits
      .map((commit) => commit.message)
      .join("*");
  }

  return "no release notes";
};

const getAuthor = (githubContext: GithubContextType): string =>
  githubContext.event.release &&
  !githubContext.event.release.author.login.includes("github-actions")
    ? githubContext.event.release.author.login
    : githubContext.actor;

const getDateTime = (): string => {
  return Intl.DateTimeFormat("en", {
    timeZone: "America/Chicago",
    day: "2-digit",
    month: "2-digit",
    year: "numeric",
    hour: "numeric",
    minute: "numeric",
    second: "numeric",
    timeZoneName: "short",
  })
    .format(new Date())
    .replace(/\u202F/g, " ");
};

const getStartedTimestamp = (): string => {
  if (process.env.STARTED_TIMESTAMP) {
    return process.env.STARTED_TIMESTAMP;
  }
  exportVariable("STARTED_TIMESTAMP", getDateTime());

  return getDateTime();
};

const getStoppedTimestamp = (): string => getDateTime();

const useConfig = (): ConfigType => {
  console.log(JSON.stringify(JSON.parse(getInput("github")), null, 2));
  const githubContext: GithubContextType = JSON.parse(getInput("github"));
  const status = getInput("status");
  const type = getInput("type");
  const channel = getInput("channel");
  const mention_person = getInput("mention-person");
  const job_status = process.env.job_status;
  const startedTimestamp = getStartedTimestamp();
  const stoppedTimestamp = getStoppedTimestamp();

  return {
    repository: githubContext.event.repository.name,
    version: getVersion(githubContext),
    release_notes: getReleaseNotes(githubContext),
    author: getAuthor(githubContext),
    action_url: `${githubContext.server_url}/${githubContext.repository}/actions/runs/${githubContext.run_id}`,
    message_id: process.env.SLACK_MESSAGE_ID,
    startedTimestamp,
    stoppedTimestamp,
    status,
    type,
    job_status,
    channel,
    mention_person,
  };
};

export default useConfig;
