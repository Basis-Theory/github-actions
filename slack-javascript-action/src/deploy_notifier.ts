import useConfig from "./useConfig";
import {
  approvalWasGrantedOrRejected,
  askForApproval,
} from "./approval.helpers";
import {
  alertDeployDone,
  alertDeployStarting,
  threadReleaseNotes,
} from "./deploy.helpers";
import { draftReleaseIsReady } from "./draft_release_ready.helpers";
import { COMPLETED_STATUSES } from "./github.helpers";

export const deploy_notifier = async () => {
  const config = useConfig();

  if (config.type === "draft-release-ready") {
    return await draftReleaseIsReady(config);
  } else {
    if (config.status === "request") {
      return await askForApproval(config);
    } else if (
      COMPLETED_STATUSES.includes(config.status) ||
      config.message_id
    ) {
      return await alertDeployDone(config);
    } else {
      const message = await alertDeployStarting(config);
      const releaseNotes = await threadReleaseNotes({
        ...config,
        message_id: message.ts,
      });
      const approvalGranted = await approvalWasGrantedOrRejected(
        config,
        message
      );

      return { message, releaseNotes, approvalGranted };
    }
  }
};
