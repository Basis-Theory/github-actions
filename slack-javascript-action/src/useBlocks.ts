import { ConfigType } from "./useConfig";
import { SlackMessage } from "./slack.client";

const releaseNotesToBlocks = (release_notes: string): any => {
  const fullChangelogRegex = /Full Changelog.*/i;
  release_notes = release_notes
    .replace("## What's Changed", "")
    .replace(fullChangelogRegex, "")
    .replace("\r\n", "")
    .trim();

  if (release_notes === "") return [];

  const note_blocks = release_notes
    .split("*")
    .filter((item) => item !== "")
    .map((item) => {
      const urlRegex = /(.*) in (.*)/i;
      const match = urlRegex.exec(item);
      let note = !match ? item.trim() : match[1].trim();

      return {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `â€¢ ${note}`,
        },
      };
    });

  return [
    {
      type: "section",
      text: {
        type: "mrkdwn",
        text: " :loud_sound: *Release Notes* :loud_sound:",
      },
    },
    ...note_blocks,
  ];
};

const getApprovalMessage = (
  { repository, version, author, action_url, mention_person }: ConfigType,
  release_message: SlackMessage | undefined = undefined,
  completed: boolean = false
): any => {
  let header_text = `${completed ? ":approved: ~" : ""}*Approval Requested`;
  header_text += mention_person ? ` from <${mention_person}>*` : "*";
  header_text += completed ? `~` : "";

  return [
    {
      type: "section",
      text: {
        type: "mrkdwn",
        text: header_text,
      },
    },
    {
      type: "context",
      elements: [
        {
          text: `${
            completed ? "~" : ""
          }:git: \`${repository}\` @ \`${version}\`  | :technologist: ${author}${
            completed ? "~" : ""
          }`,
          type: "mrkdwn",
        },
      ],
    },
    {
      type: "actions",
      elements: [
        {
          type: "button",
          text: {
            type: "plain_text",
            text: completed ? "Open Deploy :slack:" : "Open Action :github:",
            emoji: true,
          },
          url: completed
            ? `https://basistheory.slack.com/archives/${release_message?.channel}/${release_message?.ts}`
            : action_url,
        },
      ],
    },
    {
      type: "divider",
    },
  ];
};

const getDraftReleaseReadyMessage = ({
  repository,
  version,
}: ConfigType): any => {
  let header_text = `:package: New Draft Version Created: ${repository}@${version}`;

  return [
    {
      type: "header",
      text: {
        type: "plain_text",
        text: header_text,
      },
    },
    {
      type: "actions",
      elements: [
        {
          type: "button",
          text: {
            type: "plain_text",
            text: "Open Release :slack:",
            emoji: true,
          },
          url: `https://github.com/Basis-Theory/${repository}/releases/edit/${version}`,
        },
      ],
    },
    {
      type: "divider",
    },
  ];
};

const getFailedMention = ({ mention_person }: ConfigType): any => {
  const mention = mention_person ? mention_person : "!subteam^S04RC9KQ77F";
  return [
    {
      type: "section",
      text: {
        type: "mrkdwn",
        text: `<${mention}>`,
      },
    },
  ];
};

const getDeployMessage = (
  heading: string,
  {
    repository,
    version,
    author,
    action_url,
    status,
    startedTimestamp,
    stoppedTimestamp,
  }: ConfigType
): any => [
  {
    type: "header",
    text: {
      type: "plain_text",
      text: heading,
    },
  },
  {
    type: "context",
    elements: [
      {
        text: `:git: \`${repository}\` @ \`${version}\`  | :technologist: ${author}`,
        type: "mrkdwn",
      },
    ],
  },
  {
    type: "context",
    elements: [
      {
        text: `Deploy started \`${startedTimestamp}\` ${
          status === "done" ? `and finished \`${stoppedTimestamp}\`` : ""
        }`,
        type: "mrkdwn",
      },
    ],
  },
  {
    type: "actions",
    elements: [
      {
        type: "button",
        text: {
          type: "plain_text",
          text: "Open Action  :github:",
          emoji: true,
        },
        url: action_url,
      },
    ],
  },
  {
    type: "divider",
  },
];

const useBlocks = () => ({
  releaseNotesToBlocks,
  getApprovalMessage,
  getDraftReleaseReadyMessage,
  getDeployMessage,
  getFailedMention,
});

export default useBlocks;
