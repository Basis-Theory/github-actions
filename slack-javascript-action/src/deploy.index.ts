import { deploy_notifier } from "./deploy_notifier";

deploy_notifier().then().catch(console.error);
