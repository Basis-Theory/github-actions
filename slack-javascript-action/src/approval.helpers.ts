import { sendMessage, SlackMessage, updateMessage } from "./slack.client";
import useBlocks from "./useBlocks";
import { ConfigType } from "./useConfig";
import fs from "fs";
import { DefaultArtifactClient } from "@actions/artifact";
import { threadReleaseNotes } from "./deploy.helpers";

const FILE_NAME = "release-message-information";

const askForApproval = async (config: ConfigType) => {
  const message = await sendMessage(
    config.channel,
    useBlocks().getApprovalMessage(config)
  );
  const releaseNotes = await threadReleaseNotes({
    ...config,
    message_id: message.ts,
  });

  fs.writeFileSync(
    FILE_NAME,
    JSON.stringify({
      channel: config.channel,
      message_id: message.ts,
    }),
    { encoding: "utf8" }
  );
  const artifactClient = new DefaultArtifactClient();
  await artifactClient.uploadArtifact(FILE_NAME, [`./${FILE_NAME}`], ".");

  return message.ts;
};

const approvalWasGranted = async (
  config: ConfigType,
  release_message: SlackMessage
): Promise<string | undefined> => {
  const messageBlocks = useBlocks().getApprovalMessage(
    config,
    release_message,
    true
  );

  try {
    const artifactClient = new DefaultArtifactClient();
    const { artifact } = await artifactClient.getArtifact(FILE_NAME);
    if (artifact) {
      const { downloadPath } = await artifactClient.downloadArtifact(
        artifact.id
      );

      const fileLocation = `${downloadPath}/${FILE_NAME}`;
      const content = fs.readFileSync(fileLocation, {
        encoding: "utf8",
      });
      fs.unlink(fileLocation, () => {});
      const { message_id, channel } = JSON.parse(content);

      let message;
      if (message_id) {
        message = await updateMessage(channel, message_id, messageBlocks, "");
      }

      return message?.ts;
    }
  } catch (e) {
    console.info("Problem updating file:", e);
  }

  return undefined;
};

export { askForApproval, approvalWasGranted };
