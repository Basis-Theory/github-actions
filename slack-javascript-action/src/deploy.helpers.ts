import { sendMessage, updateMessage } from "./slack.client";
import * as core from "@actions/core";
import useBlocks from "./useBlocks";
import { ConfigType } from "./useConfig";

const alertDeployStarting = async (config: ConfigType) => {
  const message = await sendMessage(
    config.channel,
    useBlocks().getDeployMessage(`:rocket: Deploy Started`, config)
  );

  core.exportVariable("SLACK_MESSAGE_ID", message.ts);

  return message;
};

const getDoneHeading = (job_status: string | undefined) => {
  if (job_status === "success") {
    return `:white_check_mark: Deploy Success`;
  } else if (job_status === "failure") {
    return `:no_entry_sign: Deploy Failure`;
  }
  return `:octagonal_sign: Deploy Cancelled`;
};

const alertDeployDone = async (config: ConfigType) => {
  const { message_id, job_status, mention_person }: ConfigType = config;
  let deployMessage = useBlocks().getDeployMessage(
    getDoneHeading(job_status),
    config
  );

  let message;
  if (message_id) {
    message = await updateMessage(config.channel, message_id, deployMessage);
  } else {
    message = await sendMessage(config.channel, deployMessage);
  }

  if(job_status === "failure" && mention_person) {
    await sendMessage(config.channel, undefined, `<${mention_person}>`, message.ts);
  }

  return message.ts;
};

const threadReleaseNotes = async ({
  channel,
  message_id,
  release_notes,
}: ConfigType) => {
  const message = await sendMessage(
    channel,
    useBlocks().releaseNotesToBlocks(release_notes),
    `${release_notes}`,
    message_id
  );

  return message;
};

export { alertDeployStarting, alertDeployDone, threadReleaseNotes };
