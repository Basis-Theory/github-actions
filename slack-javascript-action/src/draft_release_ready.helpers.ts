import { sendMessage } from "./slack.client";
import useBlocks from "./useBlocks";
import { ConfigType } from "./useConfig";

const draftReleaseIsReady = async (config: ConfigType) => {
  const message = await sendMessage(
    config.channel,
    useBlocks().getDraftReleaseReadyMessage(config)
  );

  return message.ts;
};

export { draftReleaseIsReady };
