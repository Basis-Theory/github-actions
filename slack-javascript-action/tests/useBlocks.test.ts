import useBlocks from "../src/useBlocks";
import { ConfigType } from "../src/useConfig";

describe("release notes", () => {
  test("Nothing will return nothing", () => {
    const githubReleaseNotes = ``;

    const slackReleaseNotes =
      useBlocks().releaseNotesToBlocks(githubReleaseNotes);

    expect(slackReleaseNotes).toMatchSnapshot();
  });

  test("Malformed changelog will just return the note", () => {
    const githubReleaseNotes = `## What's Changed\n * stuff`;

    const slackReleaseNotes =
      useBlocks().releaseNotesToBlocks(githubReleaseNotes);

    expect(slackReleaseNotes).toMatchSnapshot();
  });

  test("No bullets anywhere will just return the note", () => {
    const githubReleaseNotes = `## What's Changed\n no bullets \n any \n where`;

    const slackReleaseNotes =
      useBlocks().releaseNotesToBlocks(githubReleaseNotes);

    expect(slackReleaseNotes).toMatchSnapshot();
  });

  test("Correctly formed will return each as its own block and no urls", () => {
    const githubReleaseNotes = `## What's Changed \n * chore(deps): Upgrades dependencies by @dhudec in https://github.com/Basis-Theory/basistheory-vault-api/pull/1389 \n * supporting sessions in proxy by @jleon15 in https://github.com/Basis-Theory/basistheory-vault-api/pull/1390`;

    const slackReleaseNotes =
      useBlocks().releaseNotesToBlocks(githubReleaseNotes);

    expect(slackReleaseNotes).toMatchSnapshot();
  });
});

describe("get approval message", () => {
  test("Approval needed request for product", () => {
    const config: ConfigType = {
      job_status: undefined,
      message_id: undefined,
      release_notes: "",
      startedTimestamp: "",
      status: "",
      stoppedTimestamp: "",
      repository: "test-repo",
      version: "v42.0.0",
      author: "drewsue",
      action_url: "http://test-repo.com/action",
      channel: "C1234567890",
      mention_person: "drewsue",
    };

    const approvalMessage = useBlocks().getApprovalMessage(config);

    expect(approvalMessage).toMatchSnapshot();
  });

  test("Approval needed request for platform", () => {
    const config: ConfigType = {
      job_status: undefined,
      message_id: undefined,
      release_notes: "",
      startedTimestamp: "",
      status: "",
      stoppedTimestamp: "",
      repository: "test-repo",
      version: "v42.0.0",
      author: "drewsue",
      action_url: "http://test-repo.com/action",
      channel: "C1234567890",
      mention_person: "drewsue",
    };

    const approvalMessage = useBlocks().getApprovalMessage(config);

    expect(approvalMessage).toMatchSnapshot();
  });

  test("Updated approval message after it has been approved", () => {
    let releaseMessage = {
      ts: "1234567890.123456",
      channel: "C1234567890",
    };

    const config: ConfigType = {
      job_status: undefined,
      message_id: undefined,
      release_notes: "",
      startedTimestamp: "",
      status: "",
      stoppedTimestamp: "",
      repository: "test-repo",
      version: "v42.0.0",
      author: "drewsue",
      action_url: "http://test-repo.com/action",
      channel: "C1234567890",
      mention_person: "drewsue",
    };

    const approvalMessage = useBlocks().getApprovalMessage(
      config,
      releaseMessage,
      true
    );

    expect(approvalMessage).toMatchSnapshot();
  });
});

describe("get deploy message", () => {
  test("Deploying started message", () => {
    const config: ConfigType = {
      job_status: undefined,
      message_id: undefined,
      release_notes: "",
      status: "",
      stoppedTimestamp: "",
      repository: "test-repo",
      version: "v42.0.0",
      author: "drewsue",
      action_url: "http://test-repo.com/action",
      startedTimestamp: "2021-01-01T00:00:00Z",
      channel: "chan-123",
      mention_person: "drewsue",
    };

    const approvalMessage = useBlocks().getDeployMessage("Deploying", config);

    expect(approvalMessage).toMatchSnapshot();
  });

  test("Updated message when deploy is done", () => {
    const config: ConfigType = {
      job_status: undefined,
      message_id: undefined,
      release_notes: "",
      repository: "test-repo",
      version: "v42.0.0",
      author: "drewsue",
      action_url: "http://test-repo.com/action",
      status: "done",
      startedTimestamp: "2021-01-01T00:00:00Z",
      stoppedTimestamp: "2021-01-01T00:00:00Z",
      channel: "chan-123",
      mention_person: "drewsue",
    };

    const approvalMessage = useBlocks().getDeployMessage("Deploying", config);

    expect(approvalMessage).toMatchSnapshot();
  });
});
