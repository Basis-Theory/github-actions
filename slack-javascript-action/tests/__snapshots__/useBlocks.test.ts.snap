// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draft release ready Nothing will return nothing 1`] = `
[
  {
    "text": {
      "text": ":package: test-repo@v42.0.0",
      "type": "plain_text",
    },
    "type": "header",
  },
  {
    "elements": [
      {
        "text": "New Draft Version Created:",
        "type": "plain_text",
      },
    ],
    "type": "context",
  },
  {
    "elements": [
      {
        "text": {
          "emoji": true,
          "text": "Open Release :github:",
          "type": "plain_text",
        },
        "type": "button",
        "url": "https://github.com/Basis-Theory/test-repo/releases/edit/v42.0.0",
      },
    ],
    "type": "actions",
  },
  {
    "type": "divider",
  },
]
`;

exports[`get approval message Approval needed request for platform 1`] = `
[
  {
    "text": {
      "text": "*Approval Requested from <drewsue>*",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  {
    "elements": [
      {
        "text": ":git: \`test-repo\` @ \`v42.0.0\`  | :technologist: drewsue",
        "type": "mrkdwn",
      },
    ],
    "type": "context",
  },
  {
    "elements": [
      {
        "text": {
          "emoji": true,
          "text": "Open Action :github:",
          "type": "plain_text",
        },
        "type": "button",
        "url": "http://test-repo.com/action",
      },
    ],
    "type": "actions",
  },
  {
    "type": "divider",
  },
]
`;

exports[`get approval message Approval needed request for product 1`] = `
[
  {
    "text": {
      "text": "*Approval Requested from <drewsue>*",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  {
    "elements": [
      {
        "text": ":git: \`test-repo\` @ \`v42.0.0\`  | :technologist: drewsue",
        "type": "mrkdwn",
      },
    ],
    "type": "context",
  },
  {
    "elements": [
      {
        "text": {
          "emoji": true,
          "text": "Open Action :github:",
          "type": "plain_text",
        },
        "type": "button",
        "url": "http://test-repo.com/action",
      },
    ],
    "type": "actions",
  },
  {
    "type": "divider",
  },
]
`;

exports[`get approval message Updated approval message after it has been approved 1`] = `
[
  {
    "text": {
      "text": ":approved: ~*Approval Requested from <drewsue>*~",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  {
    "elements": [
      {
        "text": "~:git: \`test-repo\` @ \`v42.0.0\`  | :technologist: drewsue~",
        "type": "mrkdwn",
      },
    ],
    "type": "context",
  },
  {
    "elements": [
      {
        "text": {
          "emoji": true,
          "text": "Open Deploy :slack:",
          "type": "plain_text",
        },
        "type": "button",
        "url": "https://basistheory.slack.com/archives/C1234567890/1234567890.123456",
      },
    ],
    "type": "actions",
  },
  {
    "type": "divider",
  },
]
`;

exports[`get deploy message Deploying started message 1`] = `
[
  {
    "text": {
      "text": "Deploying",
      "type": "plain_text",
    },
    "type": "header",
  },
  {
    "elements": [
      {
        "text": ":git: \`test-repo\` @ \`v42.0.0\`  | :technologist: drewsue",
        "type": "mrkdwn",
      },
    ],
    "type": "context",
  },
  {
    "elements": [
      {
        "text": "Deploy started \`2021-01-01T00:00:00Z\` ",
        "type": "mrkdwn",
      },
    ],
    "type": "context",
  },
  {
    "elements": [
      {
        "text": {
          "emoji": true,
          "text": "Open Action  :github:",
          "type": "plain_text",
        },
        "type": "button",
        "url": "http://test-repo.com/action",
      },
    ],
    "type": "actions",
  },
  {
    "type": "divider",
  },
]
`;

exports[`get deploy message Updated message when deploy is done 1`] = `
[
  {
    "text": {
      "text": "Deploying",
      "type": "plain_text",
    },
    "type": "header",
  },
  {
    "elements": [
      {
        "text": ":git: \`test-repo\` @ \`v42.0.0\`  | :technologist: drewsue",
        "type": "mrkdwn",
      },
    ],
    "type": "context",
  },
  {
    "elements": [
      {
        "text": "Deploy started \`2021-01-01T00:00:00Z\` and finished \`2021-01-01T00:00:00Z\`",
        "type": "mrkdwn",
      },
    ],
    "type": "context",
  },
  {
    "elements": [
      {
        "text": {
          "emoji": true,
          "text": "Open Action  :github:",
          "type": "plain_text",
        },
        "type": "button",
        "url": "http://test-repo.com/action",
      },
    ],
    "type": "actions",
  },
  {
    "type": "divider",
  },
]
`;

exports[`release notes Correctly formed will return each as its own block and no urls 1`] = `
[
  {
    "text": {
      "text": " :loud_sound: *Release Notes* :loud_sound:",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  {
    "text": {
      "text": "• chore(deps): Upgrades dependencies by @dhudec",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  {
    "text": {
      "text": "• supporting sessions in proxy by @jleon15",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`release notes Malformed changelog will just return the note 1`] = `
[
  {
    "text": {
      "text": " :loud_sound: *Release Notes* :loud_sound:",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  {
    "text": {
      "text": "• stuff",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`release notes No bullets anywhere will just return the note 1`] = `
[
  {
    "text": {
      "text": " :loud_sound: *Release Notes* :loud_sound:",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  {
    "text": {
      "text": "• no bullets 
 any 
 where",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`release notes Nothing will return nothing 1`] = `[]`;
